044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,174 @@
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+You can see how the state transitions in the [BrokerLifecycleManager](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/BrokerLifecycleManager.scala#L379), specifically that 
+the broker transitions from STARTING to RECOVERY once it has caught up with the cluster metadata, then it transitions to RUNNING as long as it is not fenced.
+
+The broker nodes cannot catch up with the cluster metadata until the controller quorum has been formed, so they will not move out of the STARTING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to join the controller quorum.
+* A broker is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+* A combined node is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+        * This implies that the controller quorum will have been formed while the nodes are not marked as ready, since it cannot start accepting producer/consumer requests until it can talk to the leader.
+        * This relies on the node still accepting incoming connections from other controller nodes even if it isn't actually marked as "ready" yet.
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: controller is listening on the port of the first address in `controller.listener.names` (9090)
+* Readiness: controller is listening on the port of the first address in `controller.listener.names` (9090)
+
+### Broker only mode
+
+The proposed probes are:
+
+* Liveness: broker is listening on the port of the address of `inter.broker.listener.name` (9091)
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the brokers will not become ready until a majority of the controllers 
+are up and running.
+This is similar to the current behaviour of brokers when ZooKeeper is not ready.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: node is listening on the port of the first address in `controller.listener.names` (9090)
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
+This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 
+means the nodes will be able to communicate even if they are currently marked as not ready.
```

Comment:
The part `is listening on the port of the first address in` in this section is a bit confusing. Neither `controller.listener.names` nor `inter.broker.listener.name` contain any address the broker is listening to. They contain the listener name. It is also not configurable anyway - neither the name and the port.

So this leaves me a bit confused. Does the way it is specified here mean some sophisticated way to parse out the listener address fro the configuration? Or is it really just an overcomplicated thing to say that you will grep netstat output for listening on port 9090 / 9091? If the second, please simplify it to make it a bit more clear what does it mean.

Reply:
This was a change that was proposed by @tombentley, originally it just stated the port directly as you suggest. My assumption was that in reality the implementation would just check for port 9090/9091, however I thought it was useful in the proposal to be specific that those ports actually represent e.g. the port for the controller listener, or the port for the inter broker listener.

Reply:
I think it sounds weird and confusing this way. First, there is no address, and second, it might be read as that you do some parsing of the configs. Maybe you can then say something like this?

```
* Liveness: node is listening on the port 9090 which is configured in `controller.listener.names`
```

or something like that?

Reply:
Good suggestion, I've updated it